8307068: store a JavaThread* in the java.lang.Thread object after the JavaThread* is added to the main ThreadsList

dcubed-ojdk opened the pull request:
A small fix to store a JavaThread* in the java.lang.Thread object after the JavaThread*
is added to the main ThreadsList. Also adds the following:
- `java_lang_Thread::thread_acquire()`
- `java_lang_Thread::release_set_thread()`

And uses them on an as-needed basis.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307068](https://bugs.openjdk.org/browse/JDK-8307068): store a JavaThread* in the java.lang.Thread object after the JavaThread* is added to the main ThreadsList


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [6e6dd0a3](https://git.openjdk.org/jdk/pull/13723/files/6e6dd0a3d505db5a6a65755ed7b9234d5a2b04f1)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [6e6dd0a3](https://git.openjdk.org/jdk/pull/13723/files/6e6dd0a3d505db5a6a65755ed7b9234d5a2b04f1)
 * [Andrei Pangin](https://openjdk.org/census#apangin) (@apangin - no project role) ⚠️ Review applies to [6e6dd0a3](https://git.openjdk.org/jdk/pull/13723/files/6e6dd0a3d505db5a6a65755ed7b9234d5a2b04f1)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13723/head:pull/13723` \
`$ git checkout pull/13723`

Update a local copy of the PR: \
`$ git checkout pull/13723` \
`$ git pull https://git.openjdk.org/jdk.git pull/13723/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13723`

View PR using the GUI difftool: \
`$ git pr show -t 13723`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13723.diff">https://git.openjdk.org/jdk/pull/13723.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13723#issuecomment-1528009075)

dcubed-ojdk commented:
/label add hotspot-runtime

dholmes-ora reviewed (commented):
The eetop field is defined as a Java volatile and so the VM should nominally always perform "volatile" reads and writes ie acquires and releases. But I said nominally, because in many cases there are other synchronization "release" barriers that must have occurred before we could get to the operations that read the eetop field. So I think it is only code that could interact with just started threads, that would need the explicit acquire; while I think any set of the eetop to non-null needs to have release semantics.

dcubed-ojdk commented on src/hotspot/share/runtime/javaThread.cpp:
```diff
@@ -1672,6 +1670,10 @@ void JavaThread::prepare(jobject jni_thread, ThreadPriority prio) {
   Threads::add(this);
+  // Publish the JavaThread* in java.lang.Thread after the JavaThread* is
+  // on a ThreadsList. There will be a release when the Theads_lock is
+  // dropped somewhere in the caller, but let's be more proactive.
```
Okay. I've updated the comment to:
```
  // Publish the JavaThread* in java.lang.Thread after the JavaThread* is
  // on a ThreadsList. We don't want to wait for the release when the
  // Theads_lock is dropped somewhere in the caller since the JavaThread*
  // is already visible to JVM/TI via the ThreadsList.
```

dcubed-ojdk commented on src/hotspot/share/runtime/javaThread.cpp:
```diff
@@ -2109,6 +2108,9 @@ void JavaThread::start_internal_daemon(JavaThread* current, JavaThread* target,
   Threads::add(target); // target is now visible for safepoint/handshake
+  // Publish the JavaThread* in java.lang.Thread after the JavaThread* is
+  // on a ThreadsList.
+  java_lang_Thread::set_thread(thread_oop(), target); // isAlive == true now
```
Because the `Threads_lock` is grabbed in this function and released again
when the `mu` destructor runs, but I'm going to change my mind.  I've changed
the comment and code to:
```
  // Publish the JavaThread* in java.lang.Thread after the JavaThread* is
  // on a ThreadsList. We don't want to wait for the release when the
  // Theads_lock is dropped when the 'mu' destructor is run since the
  // JavaThread* is already visible to JVM/TI via the ThreadsList.
  java_lang_Thread::release_set_thread(thread_oop(), target); // isAlive == true now
```

dcubed-ojdk commented:
@dholmes-ora - Thanks for the review. I updated the comments and switched
from `java_lang_Thread::set_thread()` to `release_set_thread()` in the second
location.

dcubed-ojdk commented:
This fix was tested with Mach5 Tier[1-8] when it was part of:
https://github.com/openjdk/jdk/pull/13519

and is also being tested with Mach5 Tier[1-8] combined with
the fix from JDK-8307067 and what remains in JDK-8305670.

dcubed-ojdk commented:
I suppose there isn't any way to give @robehn credit for his reviews when this
fix was part of https://github.com/openjdk/jdk/pull/13519.

dholmes-ora reviewed (approved):
Updates look good and have a nice consistency! Thanks.

dcubed-ojdk commented:
@dholmes-ora - thanks for the review.

Mach5 Tier[1-8] testing of v05 results look great.

dcubed-ojdk commented:
@robehn - Any chance you are around for a re-review?
@coleenp - Any chance you can review this part since you already reviewed JDK-8307067?
@apangin - You looked at this fix when it was part of https://github.com/openjdk/jdk/pull/13519. Any chance you can take another look?

robehn reviewed (approved):
Looks good, thanks!

apangin commented:
Looks good. I only wonder if we can always use `release_set_thread` and remove plain `set_thread` entirely to avoid confusion.

dcubed-ojdk commented:
@robehn and @apangin - thanks for the re-reviews!

@apangin:
>  I only wonder if we can always use release_set_thread and remove plain set_thread entirely to avoid confusion.

I think I would prefer to leave that to another PR instead of this one.

dcubed-ojdk commented:
/integrate
